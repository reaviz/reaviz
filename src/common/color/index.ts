export * from './helper';
export * from './schemes';
