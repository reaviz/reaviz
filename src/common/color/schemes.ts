import chroma from 'chroma-js';

/**
 * Color Schemes
 * Credits: https://gka.github.io/chroma.js/#chroma-brewer
 */
export const schemes = {
  unifyviz: [
    '#4C86FF',
    '#40D3F4',
    '#40E5D1',
    '#9152EE',
    '#A840E8',
    '#80CE5B',
    '#AADC40',
    '#D740BE',
    '#EE4094',
    '#E84045',
    '#F8A340',
    '#FFD440'
  ],
  unifyvizwarm: ['#FFD440', '#F8A340', '#E84045'],
  unify8Colors: chroma
    .scale(['#4C86FF', '#26efb5'])
    .correctLightness()
    .colors(8),
  cybertron: chroma.scale(['#2d60e8', '#26efb5']).correctLightness().colors(8),
  ...chroma.brewer
};
