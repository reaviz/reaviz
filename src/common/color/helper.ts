import { extent, maxIndex } from 'd3-array';
import { scaleOrdinal, scaleQuantile } from 'd3-scale';

import { uniqueBy } from '../utils';
import { schemes } from './schemes';

export type ColorSchemeType =
  | ((data, index: number, active?: any[]) => string)
  | string[]
  | string;

export type ColorSchemeStyleArray = Partial<CSSStyleDeclaration>[];

type ColorHelperProps = {
  colorScheme: ColorSchemeType;
  point: any;
  data: any[];
  index: number;
  active: any[];
  scale?: any;
  domain?: any[];
  key?: any;
  attribute?: string;
  isMultiSeries?: boolean;
};

type ColorSchemeValueScale = (point: any) => string;

function isColorSchemeStyleArray(
  colorScheme: any
): colorScheme is ColorSchemeStyleArray {
  return Array.isArray(colorScheme) && typeof colorScheme[0] === 'object';
}

/**
 * Given a point, get the key attributes for it.
 */
const rangeHelper = (point: any, attribute: string) =>
  point.map((r, i) => {
    if (r) {
      if (r[attribute] !== undefined) {
        return r[attribute];
      } else if (r.data && r.data[attribute] !== undefined) {
        return r.data[attribute];
      }
    }

    return i;
  });

/**
 * Get a color given a range.
 */
export const getColor = (props: Partial<ColorHelperProps>) => {
  let {
    point,
    colorScheme,
    attribute,
    index,
    data,
    active,
    isMultiSeries,
    domain,
    key,
    scale
  } = {
    attribute: 'key',
    isMultiSeries: false,
    scale: scaleOrdinal,
    ...props
  };

  if (typeof colorScheme === 'string' && schemes[colorScheme]) {
    colorScheme = schemes[colorScheme];
  }

  if (Array.isArray(colorScheme)) {
    if (!domain) {
      if (isMultiSeries && Array.isArray(data)) {
        const maxIdx = maxIndex(data, (d) => d.data.length);
        const maxVal = data[maxIdx];
        data = maxVal.data;
      }

      domain = rangeHelper(data, attribute);
    }

    key = key !== undefined ? key : point[attribute];

    return scale(colorScheme).domain(domain)(key);
  } else if (typeof colorScheme === 'function') {
    return colorScheme(point, index!, active);
  } else {
    return colorScheme;
  }
};

/**
 * This function creates a value scale that maps data points to colors or CSS styles.
 *
 * @param {Array} data - The data used to create the scale.
 * @param {ColorSchemeType} colorScheme - The color scheme used to generate the scale.
 * @param {string} emptyColor - The color used for data points with no value.
 * @param {any} selections - Selected values in active state
 *
 * @returns {ColorSchemeValueScale} A function that takes a data point and returns a color or CSS style based on the data point's value.
 */
const getValueScale = (
  data,
  colorScheme: ColorSchemeType,
  emptyColor: string,
  selections: any
): ColorSchemeValueScale => {
  const valueDomain = extent(
    uniqueBy(
      data,
      (d) => d.data,
      (d) => d.value
    )
  );

  return (point) => {
    // For 0 values, lets show a placeholder fill
    if (point?.value === undefined || point?.value === null) {
      return emptyColor;
    }

    // Note: this can return css style values, not just colors
    return getColor({
      scale: scaleQuantile,
      domain: valueDomain,
      key: point.value,
      colorScheme,
      point,
      active: selections
    });
  };
};

/**
 * This function generates a style object for a given data point based on a map of value scales.
 *
 * @param {any} point - The data point for which to generate the style object.
 * @param {Map<string, ColorSchemeValueScale>} valueScales - A map where each key is a property name and each value is a function that takes a data point and returns a CSS style.
 *
 * @returns {Partial<CSSStyleDeclaration>} A style object where each key is a property name and each value is a color or CSS style based on the value of the data point for that property.
 */
export const getColorSchemeStyles = (
  point,
  valueScales: Map<string, ColorSchemeValueScale>
): Partial<CSSStyleDeclaration> =>
  Array.from(valueScales).reduce((acc, [key, valueScale]) => {
    return { ...acc, [key]: valueScale(point) };
  }, {});

/**
 * This function retrieves a color scheme for a specific property from a given color scheme.
 *
 * @param {ColorSchemeStyleArray} colorScheme - The color scheme used to generate the scale.
 * @param {string} colorSchemeProperty - The property for which to retrieve the new color scheme.
 *
 * @returns {ColorSchemeType} A color scheme for the specified property.
 */
const getColorSchemeForProperty = (
  colorScheme: ColorSchemeStyleArray,
  colorSchemeProperty: string
): ColorSchemeType =>
  colorScheme.map(
    (schemeItem: Partial<CSSStyleDeclaration>) =>
      schemeItem?.[colorSchemeProperty]
  );

/**
 * This function creates a map of value scales for different properties based on a provided color scheme.
 * Each value scale is a function that takes a data point and returns a css style value based on that point.
 *
 * @param {Array} data - The data used to create the scales.
 * @param {ColorSchemeType | ColorSchemeStyleArray} colorScheme - The color scheme used to generate the scales. This can be an array of colors or an array of objects where each object contains a set of css styles.
 * @param {string} emptyColor - The color used for data points with no value.
 * @param {any} selections - Selected values in active state
 *
 * @returns {Map<string, ColorSchemeValueScale>} A map where each key is a property name and each value is a function that takes a data point and returns a value for the property.
 *
 * If the color scheme is an array of strings, they will be treated as fill values.
 */
export const createColorSchemeValueScales = (
  data,
  colorScheme: ColorSchemeType | ColorSchemeStyleArray,
  emptyColor: string,
  selections: any
): Map<string, ColorSchemeValueScale> => {
  const valueScales = new Map<string, ColorSchemeValueScale>();

  if (isColorSchemeStyleArray(colorScheme)) {
    const colorSchemeProperties = [
      ...new Set(colorScheme.flatMap(Object.keys))
    ];

    colorSchemeProperties.forEach((key) => {
      const valueScale = getValueScale(
        data,
        getColorSchemeForProperty(colorScheme, key),
        emptyColor,
        selections
      );
      valueScales.set(key, valueScale);
    });
  } else {
    valueScales.set(
      'fill',
      getValueScale(data, colorScheme, emptyColor, selections)
    );
  }

  return valueScales;
};
