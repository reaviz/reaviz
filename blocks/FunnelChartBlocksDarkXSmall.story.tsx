import React from 'react';
import {
  FunnelArc,
  FunnelAxis,
  FunnelAxisLine,
  FunnelChart,
  FunnelSeries,
  TooltipArea
} from '../src/index';
import { largeFunnelData, simpleFunnelData } from 'reaviz-data-utils';
import { schemes } from '../src/common/color/schemes';

export default {
  tags: ['snapshot'],
  title: 'Blocks/Funnel Chart/Dark/X-Small',
  component: FunnelChart
};

export const LargeDataset = () => {
  return (
    <div className="flex flex-col pt-4 pb-4 bg-black rounded-3xl shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] w-[200px] h-[386px] overflow-hidden">
      <h3 className="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
        Incident Report
      </h3>
      <FunnelChart
        id="large-dataset"
        data={largeFunnelData}
        series={
          <FunnelSeries
            arc={
              <FunnelArc
                colorScheme={schemes.unifyvizwarm[2]}
                gradient={null}
                tooltip={<TooltipArea />}
                glow={{ blur: 20, color: '#E8404599' }}
              />
            }
            axis={
              <FunnelAxis
                label={null}
                line={<FunnelAxisLine strokeColor={'#000000'} />}
              />
            }
          />
        }
      />
    </div>
  );
};

export const Simple = () => {
  return (
    <div className="flex flex-col pt-4 pb-4 bg-black rounded-3xl shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] w-[200px] h-[386px] overflow-hidden">
      <h3 className="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
        Incident Report
      </h3>
      <FunnelChart
        id="simple"
        data={simpleFunnelData}
        series={
          <FunnelSeries
            arc={
              <FunnelArc
                colorScheme={['#5B14C5']}
                gradient={null}
                tooltip={<TooltipArea />}
                glow={{ blur: 20, color: '#5B14C5' }}
              />
            }
            axis={
              <FunnelAxis
                label={null}
                line={<FunnelAxisLine strokeColor={'#000000'} />}
              />
            }
          />
        }
      />
    </div>
  );
};

export const Layered = () => {
  return (
    <div className="flex flex-col pt-4 pb-4 bg-black rounded-3xl shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] w-[200px] h-[386px] overflow-hidden">
      <h3 className="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
        Incident Report
      </h3>
      <FunnelChart
        id="layered"
        data={simpleFunnelData}
        series={
          <FunnelSeries
            arc={
              <FunnelArc
                colorScheme={['#EE409430', '#EE409475', '#EE4094']}
                gradient={null}
                tooltip={<TooltipArea />}
                variant="layered"
                glow={{
                  blur: 20,
                  color: '#EE409499'
                }}
              />
            }
            axis={
              <FunnelAxis
                label={null}
                line={<FunnelAxisLine strokeColor={'#000000'} />}
              />
            }
          />
        }
      />
    </div>
  );
};

export const Interpolation = () => {
  return (
    <div className="flex flex-col pt-4 pb-4 bg-black rounded-3xl w-[200px] h-[386px] shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] overflow-hidden">
      <h3 className="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
        Incident Report
      </h3>
      <FunnelChart
        id="interpolation"
        data={simpleFunnelData}
        series={
          <FunnelSeries
            arc={
              <FunnelArc
                colorScheme={'#40D3F4'}
                interpolation="step"
                gradient={null}
                tooltip={<TooltipArea />}
                glow={{ blur: 15, color: '#40D3F475' }}
              />
            }
            axis={
              <FunnelAxis
                label={null}
                line={<FunnelAxisLine strokeColor={'#000000'} />}
              />
            }
          />
        }
      />
    </div>
  );
};
