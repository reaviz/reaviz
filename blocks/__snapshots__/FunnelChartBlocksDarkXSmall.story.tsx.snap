// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blocks/Funnel Chart/Dark/X-Small Interpolation smoke-test 1`] = `
<div class="flex flex-col pt-4 pb-4 bg-black rounded-3xl w-[200px] h-[386px] shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] overflow-hidden">
  <h3 class="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
    Incident Report
  </h3>
  <div class="_container_1u3dt_1"
       style="height: 100%; width: 100%;"
  >
    <svg width="200"
         height="226"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <g style="transform: translate(0px, 0px);">
          <g>
            <rect height="226"
                  width="200"
                  opacity="0"
                  cursor="auto"
            >
            </rect>
            <g pointer-events="none"
               aria-label="Visited Site: 1,000, Added to Cart: 900, Initiated Checkout: 600, Purchased: 400"
               role="graphics-document"
               style="filter: drop-shadow(rgb(64, 211, 244) 0px 0px 15px);"
            >
              <path d="M0,0L25,0L25,11.3L75,11.3L75,45.2L125,45.2L125,67.8L175,67.8L175,67.8L200,67.8L200,113L175,113L175,113L125,113L125,113L75,113L75,113L25,113L25,113L0,113Z"
                    fill="#40D3F4"
                    stroke="none"
                    opacity="1"
              >
              </path>
              <path d="M0,226L25,226L25,214.7L75,214.7L75,180.8L125,180.8L125,158.2L175,158.2L175,158.2L200,158.2L200,113L175,113L175,113L125,113L125,113L75,113L75,113L25,113L25,113L0,113Z"
                    fill="#40D3F4"
                    stroke="none"
                    opacity="1"
              >
              </path>
            </g>
          </g>
        </g>
        <line x1="50"
              y1="0"
              x2="50"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="100"
              y1="0"
              x2="100"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="150"
              y1="0"
              x2="150"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Blocks/Funnel Chart/Dark/X-Small LargeDataset smoke-test 1`] = `
<div class="flex flex-col pt-4 pb-4 bg-black rounded-3xl shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] w-[200px] h-[386px] overflow-hidden">
  <h3 class="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
    Incident Report
  </h3>
  <div class="_container_1u3dt_1"
       style="height: 100%; width: 100%;"
  >
    <svg width="200"
         height="226"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <g style="transform: translate(0px, 0px);">
          <g>
            <rect height="226"
                  width="200"
                  opacity="0"
                  cursor="auto"
            >
            </rect>
            <g pointer-events="none"
               aria-label="Visited Site: 15,000, Added to Cart: 12,000, Initiated Checkout: 11,000, Purchased: 6,000, Subscribed: 5,000, Became a Member: 3,000, Upgraded to Premium: 1,000, Became a VIP: 900"
               role="graphics-document"
               style="filter: drop-shadow(rgb(232, 64, 69) 0px 0px 20px);"
            >
              <path d="M0,28.25C8.333,34.842,16.667,41.433,25,45.2C33.333,48.967,41.667,47.083,50,50.85C58.333,54.617,66.667,75.333,75,79.1C83.333,82.867,91.667,81.925,100,84.75C108.333,87.575,116.667,92.283,125,96.05C133.333,99.817,141.667,106.973,150,107.35C158.333,107.727,166.667,107.915,175,107.915C183.333,107.915,191.667,107.915,200,107.915L200,113C191.667,113,183.333,113,175,113C166.667,113,158.333,113,150,113C141.667,113,133.333,113,125,113C116.667,113,108.333,113,100,113C91.667,113,83.333,113,75,113C66.667,113,58.333,113,50,113C41.667,113,33.333,113,25,113C16.667,113,8.333,113,0,113Z"
                    fill="#E84045"
                    stroke="none"
                    opacity="1"
              >
              </path>
              <path d="M0,197.75C8.333,191.158,16.667,184.567,25,180.8C33.333,177.033,41.667,178.917,50,175.15C58.333,171.383,66.667,150.667,75,146.9C83.333,143.133,91.667,144.075,100,141.25C108.333,138.425,116.667,133.717,125,129.95C133.333,126.183,141.667,119.027,150,118.65C158.333,118.273,166.667,118.085,175,118.085C183.333,118.085,191.667,118.085,200,118.085L200,113C191.667,113,183.333,113,175,113C166.667,113,158.333,113,150,113C141.667,113,133.333,113,125,113C116.667,113,108.333,113,100,113C91.667,113,83.333,113,75,113C66.667,113,58.333,113,50,113C41.667,113,33.333,113,25,113C16.667,113,8.333,113,0,113Z"
                    fill="#E84045"
                    stroke="none"
                    opacity="1"
              >
              </path>
            </g>
          </g>
        </g>
        <line x1="25"
              y1="0"
              x2="25"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="50"
              y1="0"
              x2="50"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="75"
              y1="0"
              x2="75"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="100"
              y1="0"
              x2="100"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="125"
              y1="0"
              x2="125"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="150"
              y1="0"
              x2="150"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="175"
              y1="0"
              x2="175"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Blocks/Funnel Chart/Dark/X-Small Layered smoke-test 1`] = `
<div class="flex flex-col pt-4 pb-4 bg-black rounded-3xl shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] w-[200px] h-[386px] overflow-hidden">
  <h3 class="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
    Incident Report
  </h3>
  <div class="_container_1u3dt_1"
       style="height: 100%; width: 100%;"
  >
    <svg width="200"
         height="226"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <g style="transform: translate(0px, 0px);">
          <g>
            <rect height="226"
                  width="200"
                  opacity="0"
                  cursor="auto"
            >
            </rect>
            <g pointer-events="none"
               aria-label="Visited Site: 1,000, Added to Cart: 900, Initiated Checkout: 600, Purchased: 400"
               role="graphics-document"
               style="filter: drop-shadow(rgb(238, 64, 148) 0px 0px 20px);"
            >
              <path d="M0,0C16.667,1.883,33.333,3.767,50,11.3C66.667,18.833,83.333,35.783,100,45.2C116.667,54.617,133.333,67.8,150,67.8C166.667,67.8,183.333,67.8,200,67.8L200,113C183.333,113,166.667,113,150,113C133.333,113,116.667,113,100,113C83.333,113,66.667,113,50,113C33.333,113,16.667,113,0,113Z"
                    fill="#EE409430"
                    stroke="none"
                    opacity="1"
              >
              </path>
              <path d="M0,226C16.667,224.117,33.333,222.233,50,214.7C66.667,207.167,83.333,190.217,100,180.8C116.667,171.383,133.333,158.2,150,158.2C166.667,158.2,183.333,158.2,200,158.2L200,113C183.333,113,166.667,113,150,113C133.333,113,116.667,113,100,113C83.333,113,66.667,113,50,113C33.333,113,16.667,113,0,113Z"
                    fill="#EE409430"
                    stroke="none"
                    opacity="1"
              >
              </path>
            </g>
          </g>
        </g>
        <g style="transform: translate(0px, 28.25px);">
          <g>
            <rect height="169.5"
                  width="200"
                  opacity="0"
                  cursor="auto"
            >
            </rect>
            <g pointer-events="none"
               aria-label="Visited Site: 1,000, Added to Cart: 900, Initiated Checkout: 600, Purchased: 400"
               role="graphics-document"
               style="filter: drop-shadow(rgb(238, 64, 148) 0px 0px 20px);"
            >
              <path d="M0,0C16.667,1.413,33.333,2.825,50,8.475C66.667,14.125,83.333,26.837,100,33.9C116.667,40.962,133.333,50.85,150,50.85C166.667,50.85,183.333,50.85,200,50.85L200,84.75C183.333,84.75,166.667,84.75,150,84.75C133.333,84.75,116.667,84.75,100,84.75C83.333,84.75,66.667,84.75,50,84.75C33.333,84.75,16.667,84.75,0,84.75Z"
                    fill="#EE409475"
                    stroke="none"
                    opacity="1"
              >
              </path>
              <path d="M0,169.5C16.667,168.088,33.333,166.675,50,161.025C66.667,155.375,83.333,142.663,100,135.6C116.667,128.538,133.333,118.65,150,118.65C166.667,118.65,183.333,118.65,200,118.65L200,84.75C183.333,84.75,166.667,84.75,150,84.75C133.333,84.75,116.667,84.75,100,84.75C83.333,84.75,66.667,84.75,50,84.75C33.333,84.75,16.667,84.75,0,84.75Z"
                    fill="#EE409475"
                    stroke="none"
                    opacity="1"
              >
              </path>
            </g>
          </g>
        </g>
        <g style="transform: translate(0px, 56.5px);">
          <g>
            <rect height="113"
                  width="200"
                  opacity="0"
                  cursor="auto"
            >
            </rect>
            <g pointer-events="none"
               aria-label="Visited Site: 1,000, Added to Cart: 900, Initiated Checkout: 600, Purchased: 400"
               role="graphics-document"
               style="filter: drop-shadow(rgb(238, 64, 148) 0px 0px 20px);"
            >
              <path d="M0,0C16.667,0.942,33.333,1.883,50,5.65C66.667,9.417,83.333,17.892,100,22.6C116.667,27.308,133.333,33.9,150,33.9C166.667,33.9,183.333,33.9,200,33.9L200,56.5C183.333,56.5,166.667,56.5,150,56.5C133.333,56.5,116.667,56.5,100,56.5C83.333,56.5,66.667,56.5,50,56.5C33.333,56.5,16.667,56.5,0,56.5Z"
                    fill="#EE4094"
                    stroke="none"
                    opacity="1"
              >
              </path>
              <path d="M0,113C16.667,112.058,33.333,111.117,50,107.35C66.667,103.583,83.333,95.108,100,90.4C116.667,85.692,133.333,79.1,150,79.1C166.667,79.1,183.333,79.1,200,79.1L200,56.5C183.333,56.5,166.667,56.5,150,56.5C133.333,56.5,116.667,56.5,100,56.5C83.333,56.5,66.667,56.5,50,56.5C33.333,56.5,16.667,56.5,0,56.5Z"
                    fill="#EE4094"
                    stroke="none"
                    opacity="1"
              >
              </path>
            </g>
          </g>
        </g>
        <line x1="50"
              y1="0"
              x2="50"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="100"
              y1="0"
              x2="100"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="150"
              y1="0"
              x2="150"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Blocks/Funnel Chart/Dark/X-Small Simple smoke-test 1`] = `
<div class="flex flex-col pt-4 pb-4 bg-black rounded-3xl shadow-[11px_21px_3px_rgba(0,0,0,0.06),14px_27px_7px_rgba(0,0,0,0.10),19px_38px_14px_rgba(0,0,0,0.13),27px_54px_27px_rgba(0,0,0,0.16),39px_78px_50px_rgba(0,0,0,0.20),55px_110px_86px_rgba(0,0,0,0.26)] w-[200px] h-[386px] overflow-hidden">
  <h3 class="text-3xl text-left p-7 pt-6 pb-8 font-bold text-white">
    Incident Report
  </h3>
  <div class="_container_1u3dt_1"
       style="height: 100%; width: 100%;"
  >
    <svg width="200"
         height="226"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <g style="transform: translate(0px, 0px);">
          <g>
            <rect height="226"
                  width="200"
                  opacity="0"
                  cursor="auto"
            >
            </rect>
            <g pointer-events="none"
               aria-label="Visited Site: 1,000, Added to Cart: 900, Initiated Checkout: 600, Purchased: 400"
               role="graphics-document"
               style="filter: drop-shadow(rgb(91, 20, 197) 0px 0px 20px);"
            >
              <path d="M0,0C16.667,1.883,33.333,3.767,50,11.3C66.667,18.833,83.333,35.783,100,45.2C116.667,54.617,133.333,67.8,150,67.8C166.667,67.8,183.333,67.8,200,67.8L200,113C183.333,113,166.667,113,150,113C133.333,113,116.667,113,100,113C83.333,113,66.667,113,50,113C33.333,113,16.667,113,0,113Z"
                    fill="#5B14C5"
                    stroke="none"
                    opacity="1"
              >
              </path>
              <path d="M0,226C16.667,224.117,33.333,222.233,50,214.7C66.667,207.167,83.333,190.217,100,180.8C116.667,171.383,133.333,158.2,150,158.2C166.667,158.2,183.333,158.2,200,158.2L200,113C183.333,113,166.667,113,150,113C133.333,113,116.667,113,100,113C83.333,113,66.667,113,50,113C33.333,113,16.667,113,0,113Z"
                    fill="#5B14C5"
                    stroke="none"
                    opacity="1"
              >
              </path>
            </g>
          </g>
        </g>
        <line x1="50"
              y1="0"
              x2="50"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="100"
              y1="0"
              x2="100"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
        <line x1="150"
              y1="0"
              x2="150"
              y2="226"
              stroke="#000000"
              stroke-width="2"
              pointer-events="none"
        >
        </line>
      </g>
    </svg>
  </div>
</div>
`;
